core/src/page_id.rs

```diff
@@ -83,19 +56,79 @@ impl ChildPageIndex {
 }
 
 impl PageId {
-    /// Construct a PageId performing a bound check.
-    /// The given bytes cannot represent a PageId bigger than the biggest valid one.
-    pub fn from_bytes(bytes: [u8; 32]) -> Result<Self, InvalidPageIdBytes> {
-        let page_id = Uint::from_be_bytes(bytes);
-        if page_id > HIGHEST_42.0 {
+    /// Decode a page ID from its disambiguated representation.
+    ///
+    /// This can fall out of bounds.
+    pub fn decode(bytes: [u8; 32]) -> Result<Self, InvalidPageIdBytes> {
+        let uint = Uint::from_be_bytes(bytes);
+
+        if uint > HIGHEST_ENCODED_42 {
             return Err(InvalidPageIdBytes);
         }
-        Ok(Self(page_id))
+
+        let leading_zeros = uint.leading_zeros();
+        let bit_count = 256 - leading_zeros;
+        let sextets = (bit_count + 5) / 6;
+
+        if bit_count == 0 {
+            return Ok(ROOT_PAGE_ID);
+        }
+
+        // first we normalize the page ID: any sextet which is all zeros
```

Comment:
Probably all the following algorithm could be simplified to

```rust
        let mut limbs = ArrayVec::new();

        for _ in 0..sextets - 1 {
            uint.sub_assign(Uint::<256, 4>::from(1));

            let x = uint & Uint::from(0b111111);
            limbs.push(x.to::<u8>());

            uint.shr_assign(DEPTH);
        }

        if uint.byte(0).trailing_zeros() < 6 {
            uint.sub_assign(Uint::<256, 4>::from(1));
            limbs.push(uint.byte(0));
        }

        limbs.reverse();
```
I tested it and it passes all tests, but the page_id test suite is not extensive, so it is likely incorrect and I may be missing some pieces

Reply:
nice, this looks like it should work.

can you stack a PR on top of it?